mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+    ) -> Tuple[ActionSelectionState, Array]:
+        """Select action to take in epsilon-greedy way. Batch and agent dims are included.
+
+            Args:
+            action_selection_state: Tuple of online parameters, previous hidden state,
+                environment timestep (used to calculate epsilon) and a random key.
+            obs: The observation from the previous timestep.
+            done: The flag timestep.last() from the previous timestep.
+
+        Returns:
+            A tuple of the updated action selection state and the chosen action.
+        """
+        params, hidden_state, t, key = action_selection_state
+
+        eps = jax.numpy.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+
+        obs = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], obs)
+        done = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], done)
+
+        next_hidden_state, eps_greedy_dist = q_net.apply(params, hidden_state, (obs, done), eps)
+
+        new_key, explore_key = jax.random.split(key, 2)
+
+        action = eps_greedy_dist.sample(seed=explore_key)
+        action = jnp.squeeze(action)
+
+        next_action_selection_state = ActionSelectionState(
+            params, next_hidden_state, t + cfg.arch.num_envs, new_key
+        )
+
+        return next_action_selection_state, action
+
+    # INTERACT LEVEL 1
+    def interaction_step(
+        interaction_state: InteractionState, _: Any
+    ) -> Tuple[InteractionState, Dict]:
+        """Selects action, steps global env, stores timesteps in rb and repacks the parameters."""
+        # light unpacking
+        action_selection_state, env_state, buffer_state, obs, done = interaction_state
+
+        # select the actions to take
+        next_action_selection_state, action = select_eps_greedy_action(
+            action_selection_state, obs, done
+        )
+
+        # step env with selected actions
+        next_env_state, next_timestep = jax.vmap(env.step)(env_state, action)
+
+        # Get reward
+        reward = next_timestep.reward
+        meaned_reward = jax.tree_util.tree_map(
+            lambda x: jnp.repeat(x[:, jnp.newaxis, ...], cfg.system.num_agents, axis=1),
+            jnp.mean(next_timestep.reward, axis=-1),
+        )
+
+        reward = jax.lax.select(
+            (jnp.zeros_like(reward) + cfg.system.mean_reward).astype(int), meaned_reward, reward
+        )
+
+        transition = Transition(obs, action, reward, done)
+        transition = jax.tree_util.tree_map(
+            lambda x: x[:, jnp.newaxis, ...], transition
+        )  # Add dummy time dim
+        next_buffer_state = rb.add(buffer_state, transition)
+
+        # Next obs and done for learner state
+        next_obs = next_timestep.observation
+        next_done = next_timestep.last()[
+            ..., jnp.newaxis
+        ]  # make compatible with network input and transition storage in next step
```

Comment:
Same here about weird formatting and also don't you want to add a leading axis here?
```suggestion
        # make compatible with network input and transition storage in next step
        next_done = next_timestep.last()[..., jnp.newaxis]  
```

Reply:
Nope! The replay buffer takes in content as (B, T, ...), where the experience should have *a* dimension, so we make it (B, T, 1) from (B, T)